PUBLIC REVIEW: 0005-application-formalization.md (#29)
folio-org/rfcs: text/0005-application-formalization.md

* Start Date: 10/09/2023
* Contributors:
  * [Craig McNally](cmcnally@ebsco.com)
  * [Vince Bareau](vbareau@ebsco.com)
* RFC PRs:
  * PRELIMINARY REVIEW: https://github.com/folio-org/rfcs/pull/14
  * DRAFT REFINEMENT: https://github.com/folio-org/rfcs/pull/22
  * PUBLIC REVIEW: https://github.com/folio-org/rfcs/pull/29
  * FINAL REVIEW: TBD
* Outcome: <!-- Leave this blank. Will eventually be either ACCEPTED or REJECTED -->

# Application Formalization

## Summary
It is proposed that an Application be defined as the minimal but complete set of elements which together are intended to deliver a specific solution to Folio. The intention is to group together all the closely related components which will need to evolve together, but independently of other parts (i.e. other applications) of Folio.

## Motivation
The existing Folio project lacks a formal definition of an Application structure. A consequence is that Folio is stuck in a mode of monolithic releases which are ever increasing in size.  There have also been repeated complaints that the setup and management of Folio can be difficult due to the number modules involved.

> 

Reviewer: How does the notion of an Application address the problem of monolithic releases? Would FOLIO release become smaller because this notion is introduced? Would some existing functionality be removed from FOLIO release and instead provided in form of an external Application? If so, which specific functionality would that be?

* Application management and deployment are ***out of scope***
* Application Stores and Marketplaces are ***out of scope***
* Application-specific UI bundles is ***out of scope***
* How movement of modules between Applications will work is ***out of scope***
* How bounded contexts and cross-module database access could be introduced is ***out of scope***

## Detailed Explanation/Design

### Terminology
* Application:  The minimal but complete set of elements which together are intended to deliver a specific solution to Folio.
* Bounded context:  The boundary of the microservice.  It "owns" the storage for the service.

### Specifics

#### Application Composition
Applications can be thought of as vertical slices of functionality.  As such, they're are comprised of a combination of the following:
* Backend Modules 
* Edge Modules 
* UI Modules 
* UI Plugins 

Applications will typically be made up of multiple components spanning all or several of these categories.  However, that's not a requirement.  In fact, in it's simplest form, it's possible for an application to be comprised of a single module from any of the categories listed above.

> An important concept introduced by the formalization of Applications, is that Applications become the delivery mechanism for the modules they contain. Therefore, if a new version of a particular module is required, then it follows that the new version of the entire Application must be installed. It then also follows that a Module must belong to one and only one Application. To do otherwise is possible, but would quickly lead to dependency and version management challenges.  It is recognized that there must be accommodations to allow refactoring or consolidation of Applications, which could involve moving of particular Modules from one Application to another.

Reviewer 2: > ... It then also follows that a Module must belong to one and only one Application. To do otherwise is possible, but would quickly lead to dependency and version management challenges.

I do not think this follows, though I agree that to do otherwise has issues. Two more-specific comments:
* Without the definition of a context (say a particular installation/tenant - not defined here) then this is an unenforceable restriction unless it happens within a particular dependency chain of application descriptors (defined here).
* The current text with unspecified context (which might extend to all versions of all application descriptors) doesn't really make sense and will almost certainly be untrue as applications evolve and modules are moved about.

Because this RFC really only looks in detail at applications and modules I think it might be best to state a looser expectation that could be refined in the remaining RFCs for application formalization that consider composing applications for a tenant. Maybe something like:
> It is expected that a Module will belong to one and only one Application within a given tenant configuration.

Author: Do the following adjustments help?
> It then also follows that a **given version of a** Module must belong to one and only one Application. To do otherwise is possible, but would quickly lead to dependency and version management challenges.  To do otherwise is possible, but would quickly lead to dependency and version management challenges.  **This restriction is enforced by the Application Manager and/or Folio Application Registry (FAR), both of which are out of scope for this RFC.**  It is recognized that there must be accommodations to allow refactoring or consolidation of Applications, which could involve moving of particular Modules from one Application to another.

Reviewer 2: I think you still need to specify a context for this make sense (and "follows" is not true though not so important).

Reviewer 2: I think from comments such as https://github.com/folio-org/rfcs/pull/29#discussion_r1605111491 that the intended context is the configuration for a tenant

Author: I've incorporated the language above.

#### Application Dependencies
Formalizing Applications does not change the underlying module implementations.  This means modules which depend on interfaces provided by other modules continue to have those dependencies, and those dependencies may span Application boundaries.  However, the Application formalization rolls those interface dependencies up into Application dependencies, or dependencies on specific versions of applications. 

Example w/ many version omitted for clarity:
**app-foo-1.0.1**
* requires: 
  * `app-bar: ^1.0.0`
* modules:
  * `ui-foo`
    * requires interfaces: `foo`
  * `mod-foo`
    * provides interfaces: `foo`
    * requires interfaces: `foo-storage`, `bar`
  * `mod-foo-storage`
    * provides interfaces: `foo-storage`
    * requires interfaces: 

**app-bar-1.0.0**
* requires: 
  * `app-configuration: ~1.1.1`
* modules:
  * `mod-bar`
    * provides interfaces: `bar`
    * requires interfaces: `configuration`
  * `ui-bar`
    * requires interfaces: `bar`

**app-configuration-1.1.1**
* requires: 
* modules:
  * `mod-configuration`
    * provides interfaces: `configuration`
    * requires interfaces: 

> In this scenario, in order for `app-foo-1.0.1` to be enabled for a tenant, `app-bar-1.X.X` and `app-configuraturation-1.1.X` would also need to be enabled for the tenant.  This is because `app-foo` depends on a compatible version (specified as `^`) of `app-bar-1.X.X`, and `app-bar` depends on an approximately equivalent version (specified as `~`) of `app-configuration-1.1.1`.  Exact versions and/or ranges can also be specified.  See the [node-semver documentation](https://github.com/npm/node-semver#advanced-range-syntax]) for details.

Author: You would not be able to enable both applications for a tenant.

Similar to how modules are explicitly defind through module descriptors, applications are defined through application descriptors.  The application descriptor references module descriptors

**Application Descriptor Properties**
| Property            | Type         | Description                                                                          | Notes                    |
| ------------------- | ------------ | ------------------------------------------------------------------------------------ | ------------------------ |
| id                  | String       | Identifier for the application conforming to pattern: `{name}-{semantic version}`    | e.g. "app-orders-1.0.0"  |
| name                | String       | Name of the application                                                              | e.g. "app-orders"        |
| version             | String       | Version of the application                                                           | e.g. "1.0.0"             |
| description         | String       | Brief description of the application                                                 | e.g. "Application delivering orders functionality" |

> | dependencies        | Dependency[] | List of dependencies on other applications                                           | See "Dependency Properties" table below            |

Reviewer 3: Should the optional dependencies also be specified there? For example some application A could interact with another application B. If application B is not installed, the functionality related to the B will be hidden in A.

Author: Good point @psmagin!  I'll try to get this worked into the RFC.

Author: @psmagin after thinking through this and a bit of discussion with others we can't really think of any use cases for this.  The optional dependencies are still captured in the module descriptors, which aren't going away.  Of course, if there's a good reason to do this we can revisit later.  Thanks again for the suggestion!

| url      | String  | URL pointing to the module descriptor for this module                | e.g. "https://folio-registry.dev.folio.org/_/proxy/modules/mod-orders-storage-13.5.0" |

Example:
```
{
  "id": "app-gobi-0.0.1",
  "name": "app-gobi",
  "version": "0.0.1",
  "description": "Application facilitating the placement of orders via GOBI and other compatible vendors",
  "dependencies": [
    {
      "id": "app-orders-0.0.1",
      "name": "app-orders",
      "version": "^0.0.1"
    },
    {
      "id": "app-organizations-0.0.1",
      "name": "app-organizations",
      "version": "^0.0.1"
    }
  ],
  "modules": [
    {
      "id": "mod-gobi-2.6.0",
      "name": "mod-gobi",
      "version": "2.6.0",
	  "url": "https://folio-registry.dev.folio.org/_/proxy/modules/mod-gobi-2.6.0"
    }
  ],
  "uiModules": [
    {
      "id": "folio_gobi-settings-2.0.0",
      "name": "folio_gobi-settings",
      "version": "2.0.0",
	  "url": "https://folio-registry.dev.folio.org/_/proxy/modules/folio_gobi-settings-2.0.0"
    }
  ]
}
```
**N.B.** Full module descriptors and metadata have been omitted for readability reasons.

## Benefits

> * This is the first step toward breaking the monolithic release into application-specific releases

Reviewer: This claim is repeated in the "motivation" section but there's no example of a FOLIO release based on the Application structure to support it. The RFC should expand on this, using concrete examples of existing functionality delegated to Applications, or remove this claim and address it a separate RFC about restructuring FOLIO into Applications.

Author: Would it help if we rephrased this as:
> Application formalization is the necessary first step in being able to create Application-level releases.  How Application releases are packaged into a Folio release is the subject of a future RFC about Platforms.

Author: I've incorporated the rephrasing above

> * Results in fewer things to work with at the system operator level

Reviewer: Can you expand on why this is the case? Would the operator not need to understand the composition of the Application (what modules make it up), especially that this map to runtime components (Docker containers)?

* Brings us closer to realizing the idea of having an application store/marketplace

> * Allows for introducing/expanding the bounded context (subject of a separate RFC)

Reviewer: It's unclear what this statement means and why this is a benefit.

Author: How about this?
```suggestion
* Application formalization facilitates the adoption of a formalized microservice bounded context (subject of a separate RFC: https://github.com/folio-org/rfcs/pull/20)
```

## Risks and Drawbacks

> * Folio has too many module dependencies which must be reduced

Author: Agree.  This is not a risk of adopting Application formalization.
```suggestion
```

> * Transition to Applications will need to be iterative

Reviewer: Why? Is this approach required? Why can the formalization start with smaller Applications instead?

Author: It's not required, but we see this as the most practical way to do it.
```suggestion
* Transition to Applications should be iterative
```

  * First pass is to create larger applications spanning multiple areas in order to satisfy dependencies
  * Subsequent passes will further break up these large applications into smaller, applications focused on a single area of functionality
* Since Applications are a vertical slice of functionality, better alignment/coordination between back-end and front-end development teams is required.
  * Reorganizing development teams around applications would help
* By latching onto the Application term, there may be confusion about terminology since that term currently means different things to different people.
  * It may help to proposed formal names for other things currently referred to as Applications (e.g. the icons that appear in the Folio toolbar)

## Rationale and Alternatives
Alternative approaches include:

> * **Status quo** - Don't do anything and deal with monolithic releases, an ever-growing number of modules, and all the problems and challenges associated with those things.

Author: ```suggestion
* **Status quo** - Don't do anything and deal with monolithic releases, an ever-growing number of modules, and all the problems and challenges associated with those things.
* Releasing of coarser granularity packages of modules; e.g. platform-core, platform-minimal, platform-complete, etc.
```
